Exposed APIs in System.Xml.XmlSerializer.

shmao opened the pull request:
Ref #10572, #10574, #10575, #10577 

cc: @mconnew @zhenlan @huanwu

shmao commented:
@dotnet-bot test Innerloop Windows_NT please

shmao commented:
@dotnet-bot test Innerloop Windows_NT

shmao commented:
@dotnet-bot test Innerloop Windows_NT

shmao commented:
@dotnet-bot test Innerloop Windows_NT Debug Build and Test, test Innerloop Windows_NT Release Build and Test

zhenlan commented:
LGTM

cc: @weshaggard fyi

weshaggard commented:
You need to version the contract as it doesn't look like it has been updated yet.

shmao commented:
I changed the version of the contract to 4.1.0.

weshaggard commented on src/System.Xml.XmlSerializer/ref/4.0.10/System.Xml.XmlSerializer.depproj:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
```
You no longer need these depproj's so this can be removed.

shmao replied:
Do I still need the `\4.0.10` folder in `\ref` and project.json?

weshaggard commented:
>13:12:09 /mnt/resource/j/workspace/dotnet_corefx/dev_api/centos7.1_debug_prtest/Tools/Packaging.targets(1045,5): error : System.Xml.XmlSerializer should support API version 4.1.0.0 on .NETCore,Version=v5.0/win10-x86 but System.Xml.XmlSerializer/ref/netcore50/System.Xml.XmlSerializer.dll was found to support 4.0.0.0. 

@ericstj correct me if I'm wrong but this just means we need to update the supported frameworks in the pkgproj file now. https://github.com/dotnet/corefx/blob/dev/api/src/System.Xml.XmlSerializer/pkg/System.Xml.XmlSerializer.pkgproj#L12 I think you only one netcoreapp1.1 to be in the support list as that is the only implementation that has these (I believe).

ericstj commented on src/System.Xml.XmlSerializer/ref/System.Xml.XmlSerializer.csproj:
```diff
@@ -2,10 +2,10 @@
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.0.10.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
```
Can you use the dir.props that's under 'src/System.Xml.XmlSerializer' instead?  You may need to rebase this on latest master.

weshaggard replied:
@ericstj this change is in dev/api branch which doesn't have your versioning changes yet. Did you want to try and merge those down?

shmao replied:
Hmm, the dir.props under 'src/System.Xml.XmlSerializer' does not exist in dev/api branch yet. Will there be any FI/RI from master to dev/api so that we can use that file?

ericstj replied:
I think @joperezr was doing that.  Until that happens its ok to change the projects themselves.

joperezr replied:
I haven't started doing the merge from master into dev/api, right now I'm focusing on getting our netcoreapp1.1 tests running as part of the CI. I'll do this after that.

shmao replied:
@ericstj Sounds good. I'll keep the changes in the csproj files then. It seems the only issue left with the PR is that I should remove the depproj. Do I want to keep the `\4.0.10` folder in `\ref` and project.json?

weshaggard replied:
You shouldn't need that 4.0.10 folder under ref.

ericstj commented on src/System.Xml.XmlSerializer/src/System.Xml.XmlSerializer.csproj:
```diff
@@ -4,12 +4,12 @@
     <AssemblyName>System.Xml.XmlSerializer</AssemblyName>
     <RootNamespace>System.Xml.XmlSerializer</RootNamespace>
-    <AssemblyVersion>4.0.12.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
```
Ditto on the dir.props.

ericstj commented:
> I think you only one netcoreapp1.1 to be in the support list as that is the only implementation that has these (I believe).

You do need to update the pkgproj, but a little differently.  The ref went from targeting netstandard1.3 to netstandard1.5.  NetStandard1.5 is supported by net462, netcoreapp1.0, and xamarin.  So I'd expect https://github.com/dotnet/corefx/blob/dev/api/src/System.Xml.XmlSerializer/pkg/System.Xml.XmlSerializer.pkgproj#L12 to become net462;netcoreapp1.0;$(AllXamarinFrameworks).

shmao commented:
@ericstj I removed `\ref` folder and changed the pkgproj file. The build still failed at packaging verification. I got errors as shown below,

```
D:\OSS\corefx-1\corefx\Tools\Packaging.targets(1045,5): error : System.Xml.XmlSerializer should support API version 4.0.10.0 on .NETCore,Version=v5.0/win10-x86 but System.Xml.XmlSerializer/ref/netcore50/System.Xml.XmlSerializer.dll was found to support 4.0.0.0. [D:\OSS\corefx-1\corefx\src\System.Xml.XmlSerializer\pkg\System.Xml.XmlSerializer.pkgproj]
D:\OSS\corefx-1\corefx\Tools\Packaging.targets(1045,5): error : System.Xml.XmlSerializer should support API version 4.0.10.0 on .NETCore,Version=v5.0/win10-x86-aot but System.Xml.XmlSerializer/ref/netcore50/System.Xml.XmlSerializer.dll was found to support 4.0.0.0. [D:\OSS\corefx-1\corefx\src\System.Xml.XmlSerializer\pkg\System.Xml.XmlSerializer.pkgproj]
D:\OSS\corefx-1\corefx\Tools\Packaging.targets(1045,5): error : System.Xml.XmlSerializer should support API version 4.0.10.0 on .NETCore,Version=v5.0/win10-x64 but System.Xml.XmlSerializer/ref/netcore50/System.Xml.XmlSerializer.dll was found to support 4.0.0.0. [D:\OSS\corefx-1\corefx\src\System.Xml.XmlSerializer\pkg\System.Xml.XmlSerializer.pkgproj]
D:\OSS\corefx-1\corefx\Tools\Packaging.targets(1045,5): error : System.Xml.XmlSerializer should support API version 4.0.10.0 on .NETCore,Version=v5.0/win10-x64-aot but System.Xml.XmlSerializer/ref/netcore50/System.Xml.XmlSerializer.dll was found to support 4.0.0.0. [D:\OSS\corefx-1\corefx\src\System.Xml.XmlSerializer\pkg\System.Xml.XmlSerializer.pkgproj]
D:\OSS\corefx-1\corefx\Tools\Packaging.targets(1045,5): error : System.Xml.XmlSerializer should support API version 4.0.10.0 on .NETCore,Version=v5.0/win10-arm but System.Xml.XmlSerializer/ref/netcore50/System.Xml.XmlSerializer.dll was found to support 4.0.0.0. [D:\OSS\corefx-1\corefx\src\System.Xml.XmlSerializer\pkg\System.Xml.XmlSerializer.pkgproj]
D:\OSS\corefx-1\corefx\Tools\Packaging.targets(1045,5): error : System.Xml.XmlSerializer should support API version 4.0.10.0 on .NETCore,Version=v5.0/win10-arm-aot but System.Xml.XmlSerializer/ref/netcore50/System.Xml.XmlSerializer.dll was found to support 4.0.0.0. [D:\OSS\corefx-1\corefx\src\System.Xml.XmlSerializer\pkg\System.Xml.XmlSerializer.pkgproj]
D:\OSS\corefx-1\corefx\Tools\Packaging.targets(1045,5): error : System.Xml.XmlSerializer should be supported on .NETFramework,Version=v4.6.2 but has no compile assets. [D:\OSS\corefx-1\corefx\src\System.Xml.XmlSerializer\pkg\System.Xml.XmlSerializer.pkgproj]
```

Is there any other file I should change? Thanks!

shmao commented:
The build is failing on errors like,

```
D:\OSS\corefx-1\corefx\Tools\Packaging.targets(1045,5): error : System.Xml.XmlSerializer should support API version 4.1.0.0 on .NETCore,Version=v5.0/win10-x86-aot but System.Xml.XmlSerializer/runtimes/aot/lib/netcore50/System.Xml.XmlSerializer.dll was found to support 4.0.11.0. [D:\OSS\corefx-1\corefx\src\System.Xml.XmlSerializer\pkg\System.Xml.XmlSerializer.pkgproj]
```

I suspected the error was related to https://github.com/dotnet/corefx/issues/11274 (System.Private.Xml cannot be built with TargetGroup=netcore50aot).

weshaggard commented on src/System.Xml.XmlSerializer/ref/System.Xml.XmlSerializer.csproj:
```diff
@@ -2,7 +2,7 @@
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
     <OutputType>Library</OutputType>
     <PackageTargetFramework>netstandard1.3</PackageTargetFramework>
     <NuGetTargetMoniker>.NETStandard,Version=v1.3</NuGetTargetMoniker>
```
Like most others working in dev/api you are going to need to bumpt this to 1.7.

weshaggard commented on src/System.Xml.XmlSerializer/pkg/System.Xml.XmlSerializer.pkgproj:
```diff
@@ -1,13 +1,10 @@
-      <SupportedFramework>net45;netcore45;wp8;wpa81</SupportedFramework>
-    </ProjectReference>
     <ProjectReference Include="..\ref\System.Xml.XmlSerializer.csproj">
       <SupportedFramework>net46;netcore50;netcoreapp1.0;$(AllXamarinFrameworks)</SupportedFramework>
```
You are going to have to reduce this supported framework list down to netcoreapp1.1 and possibly net463. That is the netcore50 error you are seeing as that doesn't support this version of the contract. See PR https://github.com/dotnet/corefx/pull/11226 as an example.

iamjasonp commented on src/System.Xml.XmlSerializer/pkg/System.Xml.XmlSerializer.pkgproj:
```diff
@@ -1,13 +1,10 @@
 <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <Version>4.0.12.0</Version>
+    <Version>4.1.0.0</Version>
```
Does this need to be upped to 4.3.1 now?

shmao replied:
@weshaggard Jason pointed out that the XmlSerializer package's version had been changed to [4.3.0.XXXX](https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Xml.XmlSerializer). Do we still need to bump the package's version in dev/api branch?

weshaggard replied:
Once we get the package version changes from master into dev/api this will actually just go away. We are no longer tying package versions with assembly versions. It will be 4.3.0 once we get that merge.

iamjasonp replied:
Ah ok. That was the part I missed :smile:

shmao commented:
I removed all changes related to bumping version from this PR given that there would be versioning changes merged from master to dev/api. I opened https://github.com/dotnet/corefx/issues/11319 for tracking the bumping version work before 1.2.0 release.

shmao commented:
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test please

> Build timed out (after 120 minutes).
